// node-pico/lib/player/web-audio-player.js
"use strict";

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Driver = require("pico.driver.webaudio");
var AudioContext = global.AudioContext || global.webkitAudioContext;

var WebAudioPlayer = (function () {
  function WebAudioPlayer(processor) {
    _classCallCheck(this, WebAudioPlayer);

    this._driver = new Driver();
    this._driver.setup({ context: new AudioContext(), bufferLength: 2048 });
    this._driver.processor = processor;

    this.env = "webaudio";
    this.bufferLength = this._driver.bufferLength;
    this.sampleRate = this._driver.sampleRate;
  }

  _createClass(WebAudioPlayer, [{
    key: "play",
    value: function play() {
      this._driver.start();
    }
  }, {
    key: "pause",
    value: function pause() {
      this._driver.stop();
    }
  }], [{
    key: "isEnabled",
    get: function get() {
      return !!AudioContext;
    }
  }]);

  return WebAudioPlayer;
})();

module.exports = WebAudioPlayer;
